src/services/formatting/rules.ts

```diff
@@ -780,4 +789,79 @@ namespace ts.formatting {
     function isNonNullAssertionContext(context: FormattingContext): boolean {
         return context.contextNode.kind === SyntaxKind.NonNullExpression;
     }
+
+    function isSemicolonDeletionContext(context: FormattingContext): boolean {
+        let nextTokenKind = context.nextTokenSpan.kind;
+        let nextTokenStart = context.nextTokenSpan.pos;
+        if (isTrivia(nextTokenKind)) {
+            const nextRealToken = context.nextTokenParent === context.currentTokenParent
+                ? findNextToken(
+                    context.currentTokenParent,
+                    findAncestor(context.currentTokenParent, a => !a.parent)!,
+                    context.sourceFile)
+                : context.nextTokenParent.getFirstToken(context.sourceFile);
+            if (!nextRealToken) {
+                return true;
+            }
+            nextTokenKind = nextRealToken.kind;
+            nextTokenStart = nextRealToken.getStart(context.sourceFile);
+        }
+
+        const startLine = context.sourceFile.getLineAndCharacterOfPosition(context.currentTokenSpan.pos).line;
+        const endLine = context.sourceFile.getLineAndCharacterOfPosition(nextTokenStart).line;
+        if (startLine === endLine) {
+            return nextTokenKind === SyntaxKind.CloseBraceToken
+                || nextTokenKind === SyntaxKind.EndOfFileToken;
+        }
+
+        if (nextTokenKind === SyntaxKind.SemicolonClassElement ||
+            nextTokenKind === SyntaxKind.SemicolonToken
+        ) {
+            return false;
+        }
+
+        if (context.contextNode.kind === SyntaxKind.InterfaceDeclaration ||
+            context.contextNode.kind === SyntaxKind.TypeAliasDeclaration
+        ) {
+            // Can’t remove semicolon after `foo`; it would parse as a method declaration:
+            //
+            // interface I {

```

Comment:
why doesn't this same rule apply to property declarations with no type and no initialiser? and why only to property signatures inside interfaces and types?

Reply:
Because interfaces/types can’t have property declarations, and classes/objects can’t have method signatures (or any other member that starts with an open paren, AFAICT.

Unfortunately, due to the definition of where ASI occurs, determining whether ASI occurs while parsing is really easy, but determining whether a given semicolon in an already-parsed tree _would_ have been inserted by ASI were it not explicitly included (which is what this function tries to determine) is _extremely_ wishy washy. The definition is (paraphrasing a bit) “ASI occurs just about any time the parser can’t understand the next token without inserting a semicolon first.” Which means, to truly tell whether a semicolon was necessary or not, you have to try parsing without it, and see if you get the same result. That’s why this function appears so strange, and why the chances that this misses an opportunity to delete a semicolon is _much_ higher than the chance that `isSemicolonInsertionContext` is wrong—that one essentially just enumerates the places where `parser.ts` parses an optional semicolon.